pypesto/ensemble/ensemble.py

```diff
@@ -367,6 +371,155 @@ def from_sample(
+                        **kwargs)
+
+    @staticmethod
+    def from_history(
+            result: Result,
+            cutoff: float = np.inf,
+            max_size: int = np.inf,
+            max_per_start: int = np.inf,
+            **kwargs,
+    ):
+        """Construct an ensemble from the history of an optimization.
+
+        Parameters
+        ----------
+        result:
+            A pyPESTO result that contains an optimization result
+            with history recorded.
+        cutoff:
+            Exclude parameters from the optimization if the nllh
+            is higher than the `cutoff`.
+        max_size:
+            The maximum size the ensemble should be.
+        max_per_start:
+            The maximum number of vectors to be included from a
+            single optimization start.
+
+        Returns
+        -------
+        The ensemble.
+        """
+        if not result.optimize_result.list[0].history.options['trace_record']:
+            logger.warning('The optimize result has no trace. The Ensemble '
+                           'will automatically be created through '
+                           'from_optimization().')
+            return Ensemble.from_optimization_endpoints(result=result,
+                                                        cutoff=cutoff,
+                                                        max_size=max_size,
+                                                        **kwargs)
+        x_vectors = []
+        vector_tags = []
+        x_names = result.problem.x_names
+        lb = result.problem.lb_full
+        ub = result.problem.ub_full
+
+        # calculate the number of starts whose final nllh is below cutoff
+        n_starts = 0
+        for start in result.optimize_result.list:
+            if start['fval'] > cutoff:
+                break
+            n_starts += 1
+
+        if n_starts*max_per_start > max_size:
+            logger.info(f'The number of starts that can contribute an '
+                        'ensemble vector multiplied with max_per_start '
+                        'is higher than max_size. Thus we will lower '
+                        'max_per_start. If you do not want this to '
+                        'happen consider increasing max_size to '
+                        f'{max_per_start*n_starts} or decrease cutoff.')
+            max_per_start = math.floor(max_size/n_starts)
+
+        for i in range(n_starts):
+            trace_x = \
+                result.optimize_result.list[i]['history'].get_x_trace()
+            trace_fval = \
+                result.optimize_result.list[i]['history'].get_fval_trace()
+
+            # calculate number of candidates
+            n_cand = 0
+            for iter_fval in reversed(trace_fval):
+                if iter_fval > cutoff:
+                    break
+                n_cand += 1
+
+            # calculate distance between the vectors included in ensemble:
+            dist = max(math.floor(n_cand/max_per_start), 1)
```

Comment:
This will somewhat truncate the maximum possible distance...
Maybe better:
```
indices = np.round(np.linspace(0, n_cand - 1, np.min(max_per_start, n_cand)))
x_vectors = [trace_x[ind] for ind in indices]
x_tags = [(iMS, len(trace_x) - 1 - ind) for ind in indices]
```